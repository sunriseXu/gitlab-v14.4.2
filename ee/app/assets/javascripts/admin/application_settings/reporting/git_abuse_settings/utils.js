import { parseBoolean } from '~/lib/utils/common_utils';

export const parseFormProps = ({
  maxNumberOfRepositoryDownloads,
  maxNumberOfRepositoryDownloadsWithinTimePeriod,
  gitRateLimitUsersAllowlist,
  autoBanUserOnExcessiveProjectsDownload,
}) => ({
  maxNumberOfRepositoryDownloads: parseInt(maxNumberOfRepositoryDownloads, 10),
  maxNumberOfRepositoryDownloadsWithinTimePeriod: parseInt(
    maxNumberOfRepositoryDownloadsWithinTimePeriod,
    10,
  ),
  gitRateLimitUsersAllowlist: JSON.parse(gitRateLimitUsersAllowlist),
  autoBanUserOnExcessiveProjectsDownload: parseBoolean(autoBanUserOnExcessiveProjectsDownload),
});
