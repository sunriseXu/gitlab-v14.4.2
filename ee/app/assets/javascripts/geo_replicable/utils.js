import createGqClient, { fetchPolicies } from '~/lib/graphql';

/*
  This file uses a NO_CACHE policy due to the need for Geo data to always be fresh.
  The UI this serves is used to watch the "syncing" process of items and their statuses
  will need to be constantly re-queried as the user navigates around to not mistakenly
  think the sync process is broken.
*/

export const gqClient = createGqClient(
  {},
  {
    fetchPolicy: fetchPolicies.NO_CACHE,
  },
);

/**
 * This is an alias for /api/v4/graphql that bypasses the Geo proxy,
 * so we ensure that we always hit the current node, if on a secondary.
 */
export const gqGeoClient = createGqClient(
  {},
  {
    fetchPolicy: fetchPolicies.NO_CACHE,
    path: '/api/v4/geo/graphql',
  },
);

/**
 * This is a proxy of the /api/v4/geo/graphql of a specific node,
 * so we're requesting data directly from a secondary's GraphQL endpoint.
 */
export const gqGeoClientForSecondaryNodeId = (secondaryNodeId) =>
  createGqClient(
    {},
    {
      fetchPolicy: fetchPolicies.NO_CACHE,
      path: `/api/v4/geo/node_proxy/${secondaryNodeId}/graphql`,
    },
  );

/**
 * Slight optimization, if we know the target node is also the current node, or
 * if we don't know what's the target node, we look at the current node instead,
 * by going to /api/v4/geo/graphql - an alias for /api/v4/graphql that bypasses
 * the Geo proxy (so we always hit the current node here, if on a secondary).
 *
 * @param {string} currentNodeId - current node id
 * @param {string} targetNodeId - target node id
 * @returns {Object} - GraphQL client
 */
export const getGraphqlClient = (currentNodeId, targetNodeId) => {
  if (targetNodeId !== undefined && currentNodeId !== targetNodeId) {
    return gqGeoClientForSecondaryNodeId(targetNodeId);
  }

  return gqGeoClient;
};
