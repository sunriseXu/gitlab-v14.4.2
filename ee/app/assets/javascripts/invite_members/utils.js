import { memoize } from 'lodash';
import Api from '~/api';

const fetchGroupMembers = memoize((id) => Api.groupMembers(id).then((response) => response.data));

export const fetchUserIdsFromGroup = async (groupIdToInvite) => {
  const groupMembers = await fetchGroupMembers(groupIdToInvite);

  return groupMembers.map((user) => user.id);
};
