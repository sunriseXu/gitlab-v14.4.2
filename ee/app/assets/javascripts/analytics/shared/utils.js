import { dataVizBlue500, gray300 } from '@gitlab/ui/scss_to_js/scss_variables';
import { merge, cloneDeep } from 'lodash';
import { dateFormats } from '~/analytics/shared/constants';
import dateFormat from '~/lib/dateformat';
import { convertObjectPropsToCamelCase, parseBoolean } from '~/lib/utils/common_utils';
import { capitalizeFirstCharacter } from '~/lib/utils/text_utility';

export const formattedDate = (d) => dateFormat(d, dateFormats.defaultDate);

/**
 * Creates a value stream object from a dataset. Returns null if no valueStreamId is present.
 *
 * @param {Object} dataset - The raw value stream object
 * @returns {Object} - A value stream object
 */
export const buildValueStreamFromJson = (valueStream) => {
  const { id, name, is_custom: isCustom } = valueStream ? JSON.parse(valueStream) : {};
  return id ? { id, name, isCustom } : null;
};

/**
 * Creates an array of stage objects from a json string. Returns an empty array if no stages are present.
 *
 * @param {String} stages - JSON encoded array of stages
 * @returns {Array} - An array of stage objects
 */
const buildDefaultStagesFromJSON = (stages = '') => {
  if (!stages.length) return [];
  return JSON.parse(stages);
};

/**
 * Creates a group object from a dataset. Returns null if no groupId is present.
 *
 * @param {Object} dataset - The container's dataset
 * @returns {Object} - A group object
 */
export const buildGroupFromDataset = (dataset) => {
  const { groupId, groupName, groupFullPath, groupAvatarUrl, groupParentId } = dataset;

  if (groupId) {
    return {
      id: Number(groupId),
      name: groupName,
      full_path: groupFullPath,
      avatar_url: groupAvatarUrl,
      parent_id: groupParentId,
    };
  }

  return null;
};

/**
 * Creates a project object from a dataset. Returns null if no projectId is present.
 *
 * @param {Object} dataset - The container's dataset
 * @returns {Object} - A project object
 */
export const buildProjectFromDataset = (dataset) => {
  const { projectGid, projectName, projectPathWithNamespace, projectAvatarUrl } = dataset;

  if (projectGid) {
    return {
      id: projectGid,
      name: projectName,
      path_with_namespace: projectPathWithNamespace,
      avatar_url: projectAvatarUrl,
    };
  }

  return null;
};

/**
 * Creates a new date object without time zone conversion.
 *
 * We use this method instead of `new Date(date)`.
 * `new Date(date) will assume that the date string is UTC and it
 * ant return different date depending on the user's time zone.
 *
 * @param {String} date - Date string.
 * @returns {Date} - Date object.
 */
export const toLocalDate = (date) => {
  const dateParts = date.split('-');

  return new Date(dateParts[0], dateParts[1] - 1, dateParts[2]);
};

/**
 * Creates an array of project objects from a json string. Returns null if no projects are present.
 *
 * @param {String} data - JSON encoded array of projects
 * @returns {Array} - An array of project objects
 */
const buildProjectsFromJSON = (projects = '') => {
  if (!projects.length) return [];
  return JSON.parse(projects);
};

/**
 * Builds the initial data object for Value Stream Analytics with data loaded from the backend
 *
 * @param {Object} dataset - dataset object paseed to the frontend via data-* properties
 * @returns {Object} - The initial data to load the app with
 */
export const buildCycleAnalyticsInitialData = ({
  valueStream = null,
  groupId = null,
  createdBefore = null,
  createdAfter = null,
  projects = null,
  groupName = null,
  groupFullPath = null,
  groupParentId = null,
  groupAvatarUrl = null,
  labelsPath = '',
  milestonesPath = '',
  defaultStages = null,
  stage = null,
  aggregationEnabled = false,
  aggregationLastRunAt = null,
  aggregationNextRunAt = null,
} = {}) => ({
  selectedValueStream: buildValueStreamFromJson(valueStream),
  group: groupId
    ? convertObjectPropsToCamelCase(
        buildGroupFromDataset({
          groupId,
          groupName,
          groupFullPath,
          groupAvatarUrl,
          groupParentId,
        }),
      )
    : null,
  createdBefore: createdBefore ? toLocalDate(createdBefore) : null,
  createdAfter: createdAfter ? toLocalDate(createdAfter) : null,
  selectedProjects: projects
    ? buildProjectsFromJSON(projects).map(convertObjectPropsToCamelCase)
    : null,
  labelsPath,
  milestonesPath,
  defaultStageConfig: defaultStages
    ? buildDefaultStagesFromJSON(defaultStages).map(({ name, ...rest }) => ({
        ...convertObjectPropsToCamelCase(rest),
        name: capitalizeFirstCharacter(name),
      }))
    : [],
  stage: JSON.parse(stage),
  aggregation: {
    enabled: parseBoolean(aggregationEnabled),
    lastRunAt: aggregationLastRunAt,
    nextRunAt: aggregationNextRunAt,
  },
});

/**
 * Linearly interpolates between two values
 *
 * @param {Number} valueAtT0 The value at t = 0
 * @param {Number} valueAtT1 The value at t = 1
 * @param {Number} t The current value of t
 *
 * @returns {Number} The result of the linear interpolation.
 */
const lerp = (valueAtT0, valueAtT1, t) => {
  return valueAtT0 * (1 - t) + valueAtT1 * t;
};

/**
 * Builds a second series that visually represents the "no data" (i.e. "null")
 * data points, and returns a new series Array that includes both the "null"
 * and "non-null" data sets.
 * This function returns new series data and does not modify the original instance.
 *
 * @param {Array} seriesData The lead time series data that has already been processed
 * by the `apiDataToChartSeries` function above.
 * @returns {Array} A new series Array
 */
export const buildNullSeries = (seriesData, nullSeriesTitle) => {
  const nonNullSeries = cloneDeep(seriesData[0]);

  // Loop through the series data and build a list of all the "gaps". A "gap" is
  // a section of the data set that only include `null` values. Each gap object
  // includes the start and end indices and the start and end values of the gap.
  const seriesGaps = [];
  let currentGap = null;
  nonNullSeries.data.forEach(([, value], index) => {
    if (value == null && currentGap == null) {
      currentGap = {};

      if (index > 0) {
        currentGap.startIndex = index - 1;
        const [, previousValue] = nonNullSeries.data[index - 1];
        currentGap.startValue = previousValue;
      }

      seriesGaps.push(currentGap);
    } else if (value != null && currentGap != null) {
      currentGap.endIndex = index;
      currentGap.endValue = value;
      currentGap = null;
    }
  });

  // Create a copy of the non-null series, but with all the data point values set to `null`
  const nullSeriesData = nonNullSeries.data.map(([date]) => [date, null]);

  // Render each of the gaps to the "null" series. Values are determined by linearly
  // interpolating between the start and end values.
  seriesGaps.forEach((gap) => {
    const startIndex = gap.startIndex ?? 0;
    const startValue = gap.startValue ?? gap.endValue ?? 0;
    const endIndex = gap.endIndex ?? nonNullSeries.data.length - 1;
    const endValue = gap.endValue ?? gap.startValue ?? 0;

    for (let i = startIndex; i <= endIndex; i += 1) {
      const t = (i - startIndex) / (endIndex - startIndex);
      nullSeriesData[i][1] = lerp(startValue, endValue, t);
    }
  });

  merge(nonNullSeries, {
    showSymbol: true,
    showAllSymbol: true,
    symbolSize: 8,
    lineStyle: {
      color: dataVizBlue500,
    },
    areaStyle: {
      color: dataVizBlue500,
      opacity: 0,
    },
    itemStyle: {
      color: dataVizBlue500,
    },
  });

  const nullSeries = {
    name: nullSeriesTitle,
    data: nullSeriesData,
    lineStyle: {
      type: 'dashed',
      color: gray300,
    },
    areaStyle: {
      color: 'none',
    },
    itemStyle: {
      color: gray300,
    },
  };

  return [nullSeries, nonNullSeries];
};
