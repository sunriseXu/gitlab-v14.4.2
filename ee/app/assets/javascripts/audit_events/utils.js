import { parsePikadayDate, pikadayToString } from '~/lib/utils/datetime_utility';
import { getIdFromGraphQLId } from '~/graphql_shared/utils';
import {
  AVAILABLE_TOKEN_TYPES,
  AUDIT_FILTER_CONFIGS,
  ENTITY_TYPES,
  createBlankHeader,
} from './constants';
import { parseUsername, displayUsername } from './token_utils';

export const getTypeFromEntityType = (entityType) => {
  return AUDIT_FILTER_CONFIGS.find(
    ({ entityType: configEntityType }) => configEntityType === entityType,
  )?.type;
};

export const getEntityTypeFromType = (type) => {
  return AUDIT_FILTER_CONFIGS.find(({ type: configType }) => configType === type)?.entityType;
};

export const parseAuditEventSearchQuery = ({
  created_after: createdAfter,
  created_before: createdBefore,
  entity_type: entityType,
  entity_username: entityUsername,
  author_username: authorUsername,
  ...restOfParams
}) => ({
  ...restOfParams,
  created_after: createdAfter ? parsePikadayDate(createdAfter) : null,
  created_before: createdBefore ? parsePikadayDate(createdBefore) : null,
  entity_type: getTypeFromEntityType(entityType),
  entity_username: displayUsername(entityUsername),
  author_username: displayUsername(authorUsername),
});

export const createAuditEventSearchQuery = ({ filterValue, startDate, endDate, sortBy }) => {
  const entityValue = filterValue.find((value) => AVAILABLE_TOKEN_TYPES.includes(value.type));
  const entityType = getEntityTypeFromType(entityValue?.type);
  const filterData = entityValue?.value.data;

  const params = {
    created_after: startDate ? pikadayToString(startDate) : null,
    created_before: endDate ? pikadayToString(endDate) : null,
    sort: sortBy,
    entity_type: entityType,
    entity_id: null,
    entity_username: null,
    author_username: null,
    // When changing the search parameters, we should be resetting to the first page
    page: null,
  };

  switch (entityType) {
    case ENTITY_TYPES.USER:
      params.entity_username = parseUsername(filterData);
      break;
    case ENTITY_TYPES.AUTHOR:
      params.author_username = parseUsername(filterData);
      break;
    default:
      params.entity_id = filterData;
  }

  return params;
};

export const mapItemHeadersToFormData = (item, settings = {}) => {
  const headers = item?.headers?.nodes || [];

  return (
    headers
      .map(({ id, key, value }) => ({
        ...createBlankHeader(),
        id,
        name: key,
        value,
        ...settings,
      }))
      // Sort the headers so they appear in the order they were created
      // The GraphQL endpoint returns them in the reverse order of this
      .sort((a, b) => getIdFromGraphQLId(a.id) - getIdFromGraphQLId(b.id))
  );
};

export const mapAllMutationErrors = (mutations, name) => {
  return Promise.allSettled(mutations).then((results) => {
    const rejected = results.filter((r) => r.status === 'rejected').map((r) => r.reason);

    if (rejected.length > 0) {
      throw rejected[0];
    }

    return results
      .filter((r) => r.status === 'fulfilled')
      .map((r) => r.value.data[name].errors)
      .reduce((r, errors) => r.concat(errors), [])
      .filter(Boolean);
  });
};
