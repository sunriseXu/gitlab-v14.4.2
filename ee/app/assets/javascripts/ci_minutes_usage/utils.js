import { dateToYearMonthDate, formatDateAsMonth } from '~/lib/utils/datetime_utility';

export const formatYearMonthData = (ciMinutesUsage, formatMonth = false) => {
  return ciMinutesUsage.length > 0
    ? ciMinutesUsage.map((cur) => {
        const date = new Date(cur.monthIso8601);
        const formattedDate = dateToYearMonthDate(date);

        if (formatMonth) {
          return {
            ...cur,
            ...formattedDate,
            monthName: formatDateAsMonth(date, { abbreviated: false }),
          };
        }

        return {
          ...cur,
          ...formattedDate,
        };
      })
    : [];
};

export const getUsageDataByYear = (ciMinutesUsage) => {
  const formattedData = formatYearMonthData(ciMinutesUsage);

  return formattedData.reduce((prev, cur) => {
    if (prev[cur.year] != null) {
      prev[cur.year].push(cur);
    } else {
      // eslint-disable-next-line no-param-reassign
      prev[cur.year] = [cur];
    }
    return prev;
  }, {});
};

export const getSortedYears = (usageDataByYear) => {
  return Object.keys(usageDataByYear).reverse();
};
