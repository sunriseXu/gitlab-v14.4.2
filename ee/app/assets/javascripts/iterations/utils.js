import { formatDate } from '~/lib/utils/datetime_utility';

const PERIOD_DATE_FORMAT = 'mmm d, yyyy';

/**
 * The arguments are two date strings in formatted in ISO 8601 (YYYY-MM-DD)
 *
 * @returns {string} ex. "Oct 1, 2021 - Oct 10, 2021"
 */
export function getIterationPeriod({ startDate, dueDate }) {
  const start = formatDate(startDate, PERIOD_DATE_FORMAT, true);
  const due = formatDate(dueDate, PERIOD_DATE_FORMAT, true);
  return `${start} - ${due}`;
}

/**
 * Group a list of iterations by cadence.
 *
 * @param iterations A list of iterations
 * @return {Array} A list of cadences
 */
export function groupByIterationCadences(iterations) {
  const cadences = [];
  iterations.forEach((iteration) => {
    if (!iteration.iterationCadence) {
      return;
    }
    const { title, id } = iteration.iterationCadence;
    const cadenceIteration = {
      id: iteration.id,
      title: iteration.title,
      period: getIterationPeriod(iteration),
    };
    const cadence = cadences.find((c) => c.title === title);
    if (cadence) {
      cadence.iterations.push(cadenceIteration);
    } else {
      cadences.push({ title, iterations: [cadenceIteration], id });
    }
  });
  return cadences;
}
