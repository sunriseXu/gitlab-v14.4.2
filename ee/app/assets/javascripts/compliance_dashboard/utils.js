import { convertToGraphQLIds } from '~/graphql_shared/utils';
import { TYPE_PROJECT } from '~/graphql_shared/constants';
import { formatDate, getDateInPast, pikadayToString } from '~/lib/utils/datetime_utility';
import { ISO_SHORT_FORMAT } from '~/vue_shared/constants';
import { queryToObject } from '~/lib/utils/url_utility';
import { CURRENT_DATE } from '../audit_events/constants';

export const convertProjectIdsToGraphQl = (projectIds) =>
  convertToGraphQLIds(
    TYPE_PROJECT,
    projectIds.filter((id) => Boolean(id)),
  );

export const parseViolationsQueryFilter = ({ mergedBefore, mergedAfter, projectIds }) => ({
  projectIds: projectIds ? convertProjectIdsToGraphQl(projectIds) : [],
  mergedBefore: formatDate(mergedBefore, ISO_SHORT_FORMAT, true),
  mergedAfter: formatDate(mergedAfter, ISO_SHORT_FORMAT, true),
});

export const buildDefaultFilterParams = (queryString) => ({
  mergedAfter: pikadayToString(getDateInPast(CURRENT_DATE, 30)),
  mergedBefore: pikadayToString(CURRENT_DATE),
  ...queryToObject(queryString, { gatherArrays: true }),
});
