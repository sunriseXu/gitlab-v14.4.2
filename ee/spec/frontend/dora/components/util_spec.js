import lastWeekData from 'test_fixtures/api/dora/metrics/daily_lead_time_for_changes_for_last_week.json';
import { buildNullSeries } from 'ee/analytics/shared/utils';
import {
  apiDataToChartSeries,
  seriesToAverageSeries,
  seriesToMedianSeries,
  extractTimeSeriesTooltip,
  secondsToDays,
  formatAsPercentage,
} from 'ee/dora/components/util';

const NO_DATA_MESSAGE = 'No data available';

describe('ee/dora/components/util.js', () => {
  describe('apiDataToChartSeries', () => {
    it('transforms the data from the API into data the chart component can use', () => {
      const apiData = [
        // This is the date format we expect from the API
        { value: 5, date: '2015-06-28' },

        // But we should support _any_ date format
        { value: 1, date: '2015-06-28T20:00:00.000-0400' },
        { value: 8, date: '2015-07-01T00:00:00.000Z' },
      ];

      const startDate = new Date(2015, 5, 26, 10);
      const endDate = new Date(2015, 6, 4, 10);
      const chartTitle = 'Chart title';

      const expected = [
        {
          name: chartTitle,
          data: [
            ['Jun 26', 0],
            ['Jun 27', 0],
            ['Jun 28', 5],
            ['Jun 29', 1],
            ['Jun 30', 0],
            ['Jul 1', 8],
            ['Jul 2', 0],
            ['Jul 3', 0],
          ],
        },
      ];

      expect(apiDataToChartSeries(apiData, startDate, endDate, chartTitle)).toEqual(expected);
    });
  });

  describe('lead time data', () => {
    it('returns the correct lead time chart data after all processing of the API response', () => {
      const chartData = buildNullSeries(
        apiDataToChartSeries(
          lastWeekData,
          new Date(2015, 5, 27, 10),
          new Date(2015, 6, 4, 10),
          'Lead time',
          null,
        ),
        NO_DATA_MESSAGE,
      );

      expect(chartData).toMatchSnapshot();
    });
  });

  describe('seriesToAverageSeries', () => {
    const seriesName = 'Average';

    it('returns an empty object if chart data is undefined', () => {
      const data = seriesToAverageSeries(undefined, seriesName);

      expect(data).toStrictEqual({});
    });

    it('returns an empty object if chart data is blank', () => {
      const data = seriesToAverageSeries(null, seriesName);

      expect(data).toStrictEqual({});
    });

    it('returns the correct average values', () => {
      const data = seriesToAverageSeries(
        [
          ['Jul 1', 2],
          ['Jul 2', 3],
          ['Jul 3', 4],
        ],
        seriesName,
      );

      expect(data).toStrictEqual({
        name: seriesName,
        data: [
          ['Jul 1', 3],
          ['Jul 2', 3],
          ['Jul 3', 3],
        ],
      });
    });
  });

  describe('seriesToMedianSeries', () => {
    const seriesName = 'Median';

    it('returns an empty object if chart data is undefined', () => {
      const data = seriesToMedianSeries(undefined, seriesName);

      expect(data).toStrictEqual({});
    });

    it('returns an empty object if chart data is blank', () => {
      const data = seriesToMedianSeries(null, seriesName);

      expect(data).toStrictEqual({});
    });

    it('returns the correct median values', () => {
      const data = seriesToMedianSeries(
        [
          ['Jul 1', 1],
          ['Jul 2', 3],
          ['Jul 3', 10],
        ],
        seriesName,
      );

      expect(data).toStrictEqual({
        name: seriesName,
        data: [
          ['Jul 1', 3],
          ['Jul 2', 3],
          ['Jul 3', 3],
        ],
      });
    });
  });

  describe('extractTimeSeriesTooltip', () => {
    const fakeChartTitle = 'cool-chart-title';
    const params = { seriesData: [{ data: ['Apr 7', 5328] }, { data: ['Apr 7', 4000] }, {}] };

    it('displays a humanized version of the time interval in the tooltip', () => {
      const { tooltipValue } = extractTimeSeriesTooltip(params, fakeChartTitle);

      expect(tooltipValue[0].value).toBe('1.5 hours');
      expect(tooltipValue[1].value).toBe('1.1 hours');
    });

    it('will apply a custom formatter when supplied', () => {
      const formatter = jest.fn();

      extractTimeSeriesTooltip(params, fakeChartTitle, formatter);
      expect(formatter).toHaveBeenCalledTimes(2);
    });
  });

  describe('secondsToDays', () => {
    const seconds = 151000;

    it('defaults to a single decimal', () => {
      expect(secondsToDays(seconds)).toBe('1.7');
    });

    it('will format to the specified precision', () => {
      expect(secondsToDays(seconds, 3)).toBe('1.748');
    });
  });

  describe('formatAsPercentage', () => {
    it('returns 0 if given NaN', () => {
      expect(formatAsPercentage(null)).toBe('0.0%');
      expect(formatAsPercentage('a')).toBe('0.0%');
    });

    it('formats valid values', () => {
      expect(formatAsPercentage(0.25)).toBe('25.0%');
      expect(formatAsPercentage('1.86', 0)).toBe('186%');
    });
  });
});
