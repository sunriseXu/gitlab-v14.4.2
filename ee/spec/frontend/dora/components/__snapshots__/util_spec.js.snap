// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ee/dora/components/util.js lead time data returns the correct lead time chart data after all processing of the API response 1`] = `
Array [
  Object {
    "areaStyle": Object {
      "color": "none",
    },
    "data": Array [
      Array [
        "Jun 27",
        86400,
      ],
      Array [
        "Jun 28",
        86400,
      ],
      Array [
        "Jun 29",
        86400,
      ],
      Array [
        "Jun 30",
        86400,
      ],
      Array [
        "Jul 1",
        432000,
      ],
      Array [
        "Jul 2",
        518400,
      ],
      Array [
        "Jul 3",
        604800,
      ],
    ],
    "itemStyle": Object {
      "color": "#999",
    },
    "lineStyle": Object {
      "color": "#999",
      "type": "dashed",
    },
    "name": "No data available",
  },
  Object {
    "areaStyle": Object {
      "color": "#5772ff",
      "opacity": 0,
    },
    "data": Array [
      Array [
        "Jun 27",
        null,
      ],
      Array [
        "Jun 28",
        null,
      ],
      Array [
        "Jun 29",
        null,
      ],
      Array [
        "Jun 30",
        86400,
      ],
      Array [
        "Jul 1",
        432000,
      ],
      Array [
        "Jul 2",
        null,
      ],
      Array [
        "Jul 3",
        604800,
      ],
    ],
    "itemStyle": Object {
      "color": "#5772ff",
    },
    "lineStyle": Object {
      "color": "#5772ff",
    },
    "name": "Lead time",
    "showAllSymbol": true,
    "showSymbol": true,
    "symbolSize": 8,
  },
]
`;
